pygls/workspace.py

```diff
@@ -33,6 +33,58 @@
 log = logging.getLogger(__name__)
 
 
+def position_to_rowcol(lines: List[str], position: Position) -> tuple:
+    """Convert a LSP position into a row, column pair.
+
+    This method converts the Position's character offset
+    from UTF-16 code units to UTF-32 code points.
+
+    The offset of the closing quotation mark in x="😋" is
+    - 5 in UTF-16 representation
+    - 4 in UTF-32 representation
+
+    A python application can't use the character memeber of `Position`
+    directly as per specification it is represented as a zero-based line and
+    character offset based based on a UTF-16 string representation.
+
+    All characters whose codepoint exeeds the Basic Multilingual Plane are
+    represented by 2 UTF-16 code units.
+
+    see: https://github.com/microsoft/language-server-protocol/issues/376
+    """
+    row = len(lines)
+    col = 0
+    if row > position.line:
+        row = position.line
+        col = position.character
+        for ch in lines[row][:position.character]:
+            if ord(ch) > 0xFFFF:
+                col -= 1
+    return (row, col)
+
+
+def rowcol_to_position(lines: List[str], row: int, col: int) -> Position:
+    """Convert a row, column pair into a LSP Position.
+
+    This method converts the `col` argument from UTF-32 code points to
+    to UTF-16 code units and returns a `Position` object.
+
+    A python application can't use the character memeber of `Position`
+    directly as per specification it is represented as a zero-based line and
+    character offset based based on a UTF-16 string representation.
+
+    All characters whose codepoint exeeds the Basic Multilingual Plane are
+    represented by 2 UTF-16 code units.
+    """
+    line = len(lines)
+    character = 0
+    if line > row:
+        line = row
+        character = sum(1 + int(ord(ch) > 0xFFFF) for ch in lines[line][:col])
+
+    return Position(line, character)
+
+
 class Document(object):
```

Comment:
What's also problematic is that application of these changes is not atomic. You may need to put access to or mutation of the content of the document under a `threading.Lock`, as we seem to be using a threadpool executor.

(By the way, why not use the default threadpool executor from `asyncio`?)

Reply:
You are refering to the document updates, right?

Yes, reading from stdin seems to be run in a dedicated thread as there seems to be no asyncio endpoint on Windows.

Reply:
Json RPC request messages are collected using threads (as we don't want to reject/miss them), but all _builtin_ methods are executed synnchronously.

So, `textDocument/didChange` is handled in synchronous _buildin_ [method](https://github.com/openlawlibrary/pygls/blob/master/pygls/protocol.py#L609) and unless you are trying to change document content in the _pygls_ workspace from some other method, you are safe.

> (By the way, why not use the default threadpool executor from asyncio?)

Can't remember, but adding _asyncio_ was pretty challenging, specially because it is platform dependent. I was basically learning python through this project and I am aware that code could be improved. We are glad that code quality is being increased by all your contributions.

Reply:
Any update of the workspaces content creates a new string instance. Therefore any subsequent function/method using the old instance would keep working with it safely. When done the old reference is deleted and the garbage collector should remove it from memory soon.

Furthermore the `lines` past to the functions are a return value of `splitlines()` so they are independed from the original source anyway.

So even if those functions are not atomic they should not need any `threading.Lock`s from my current understandings.